File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/onboard/PinSetUpOptionScreen.java
package opencrowd.hgc.hgcwallet.ui.onboard;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Button;
import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import javax.annotation.Nullable;

import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.DBHelper;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.modals.HGCKeyType;
import opencrowd.hgc.hgcwallet.crypto.HGCSeed;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;
import opencrowd.hgc.hgcwallet.ui.auth.AuthActivity;
import opencrowd.hgc.hgcwallet.local_auth.AuthManager;
import opencrowd.hgc.hgcwallet.local_auth.AuthType;
import opencrowd.hgc.hgcwallet.ui.auth.fingerprint.FingerprintActivityHelper;
import opencrowd.hgc.hgcwallet.ui.main.MainActivity;


public class PinSetUpOptionScreen extends Screen<PinSetUpOptionView> {

    @NonNull
    HGCSeed seed;

    @NonNull
    HGCKeyType keyType;

    @Nullable
    HGCAccountID accountID;

    public PinSetUpOptionScreen(@NonNull HGCSeed seed, @NonNull HGCKeyType keyType, @Nullable HGCAccountID accountID) {
        this.seed = seed;
        this.keyType = keyType;
        this.accountID = accountID;
    }

    @NonNull
    @Override
    protected PinSetUpOptionView createView(@NonNull Context context) {
        return new PinSetUpOptionView(context);
    }

    void setupWallet() {
        AuthManager.INSTANCE.saveSeed(seed.getEntropy());
        Singleton.INSTANCE.setupWallet(keyType);
        if (accountID != null) {
            Account defaultAccount = DBHelper.getAllAccounts().get(0);
            defaultAccount.setAccountID(accountID);
            DBHelper.saveAccount(defaultAccount);
        }
        Intent intent = new Intent(getActivity(), MainActivity.class);
        getActivity().startActivity(intent);
        getActivity().finish();
    }

}

class   PinSetUpOptionView extends BaseScreenView<PinSetUpOptionScreen> {

    public PinSetUpOptionView(@NonNull Context context) {
        super(context);
        inflate(context, R.layout.view_pinsetup_layout, this);
        Button pinButton = (Button)findViewById(R.id.btn_setup_pin);
        pinButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                AuthActivity authActivity = (AuthActivity) getScreen().getActivity();
                authActivity.setupAuth(AuthType.PIN);
            }
        });

        Button deviceButton = (Button)findViewById(R.id.btn_enable_fingerprint);
        deviceButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                FingerprintActivityHelper helper = new FingerprintActivityHelper();
                helper.setup(getContext(), true, null);
                if (helper.errorMsg == null) {
                    AuthActivity authActivity = (AuthActivity) getScreen().getActivity();
                    authActivity.setupAuth(AuthType.FINGER);

                } else {
                    Singleton.INSTANCE.showToast(getScreen().getActivity(),helper.errorMsg);
                }

            }
        });
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/onboard/PinSetUpOptionScreen.kt
package hedera.hgc.hgcwallet.ui.onboard

import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.Button
import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen
import hedera.hgc.hgcwallet.Config
import hedera.hgc.hgcwallet.R
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.crypto.HGCSeed
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.local_auth.AuthManager
import hedera.hgc.hgcwallet.local_auth.AuthType
import hedera.hgc.hgcwallet.modals.HGCAccountID
import hedera.hgc.hgcwallet.modals.KeyDerivation
import hedera.hgc.hgcwallet.ui.auth.AuthActivity
import hedera.hgc.hgcwallet.ui.auth.fingerprint.FingerprintActivityHelper
import hedera.hgc.hgcwallet.ui.main.MainActivity

class PinSetUpOptionScreen(seed: HGCSeed, keyDerivation: KeyDerivation, accountId: HGCAccountID?) : Screen<PinSetUpOptionView>() {

    data class Param(val keyDerivation: KeyDerivation, val seed: HGCSeed, val accountId: HGCAccountID?)

    private val param: Param

    init {
        param = Param(keyDerivation, seed, accountId)
    }

    override fun createView(context: Context): PinSetUpOptionView {
        return PinSetUpOptionView(context, param)
    }

    fun setupWallet() {
        AuthManager.saveSeed(param.seed.getEntropy())
        Singleton.setupWallet(param.keyDerivation)
        param.accountId?.let {
            val defaultAccount = DBHelper.getAllAccounts()[0]
            defaultAccount.setAccountID(it)
            DBHelper.saveAccount(defaultAccount)
        }

        activity?.let {
            val intent = Intent(it, MainActivity::class.java)
            it.startActivity(intent)
            it.finish()
        }

    }

    internal fun onSetupButtonClick() {
        (activity as? AuthActivity)?.run {
            setupAuth(AuthType.PIN)
        }
    }

    internal fun onEnableFingerPrintClick() {
        FingerprintActivityHelper().run {
            setup(activity, true, null)
            if (errorMsg == null)
                (activity as? AuthActivity)?.run {
                    setupAuth(AuthType.FINGER)
                }
            else
                Singleton.showToast(activity, errorMsg)

        }
    }
}

class PinSetUpOptionView(context: Context, param: PinSetUpOptionScreen.Param) : BaseScreenView<PinSetUpOptionScreen>(context) {
    init {
        View.inflate(context, R.layout.view_pinsetup_layout, this)


        findViewById<Button>(R.id.btn_setup_pin)?.apply {
            text = context.getString(R.string.btn_text_setup_pin, Config.passcodeLength.toString())
            setOnClickListener {
                screen?.onSetupButtonClick()
            }

        }

        findViewById<Button>(R.id.btn_enable_fingerprint)?.apply {
            setOnClickListener {
                screen?.onEnableFingerPrintClick()
            }
        }

    }
}